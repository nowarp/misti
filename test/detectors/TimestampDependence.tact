struct Data {
    time: Int;
    value: Int;
}

contract SimpleTimestampTest {
    timestamp: Int;
    data: Data;

    init() {
        self.timestamp = now();  //BAD: Assignment uses now()
        self.data = Data {
            time: now(),         //BAD: Struct initialization with now()
            value: 0
        };
    }

    // Test arithmetic with tainted fields
    receive("test_arithmetic") {
        let temp: Int = self.data.time * 2;  //BAD: Tainted multiplication
        let sum: Int = self.timestamp + self.data.time;  //BAD: Tainted addition

        if (temp > 0) {  //BAD: Condition depends on tainted temp
            self.data.value = 1;
        }

        if (sum < 100) {  //BAD: Condition depends on tainted sum
            self.timestamp = sum;  //BAD: Propagating taint
        }

        let compound: Int = (self.data.time * 2) + (self.timestamp / 3);  //BAD: Compound tainted operation
        if (compound != 0) {  //BAD: Condition depends on compound
            self.data.value += 1;
        }
    }

    // Test variable declaration
    receive("test_variable") {
        let time: Int = now();  //BAD: Variable declaration with now()
        self.timestamp = time; //BAD: Assigning tainted variable
    }

    // Test loops
    receive("test_loops") {
        let start: Int = now();  //BAD: Variable initialization with now()

        while (now() < start + 1000) {  //BAD: Loop condition depends on now()
            self.timestamp += 1;
        }

        repeat(5) {
            self.timestamp = now();  //BAD: Repeat loop with now()
        }

        do {
            self.timestamp += 1;
        } until (now() > start + 2000);  //BAD: Until condition depends on now()
    }

    // Test returns
    fun getTimestamp(): Int {
        return now();  //BAD: Returning now() directly
    }

    // Test expressions
    receive("test_expressions") {
        let result: Int = now() > 0 ? now() : 0;  //BAD: Ternary operator with now()
        self.timestamp = result;                 //BAD: Assigning tainted result
    }
}

contract TimestampEdgeCaseTest {
    taintedValue: Int;
    timestamp: Int;

    init() {
        self.taintedValue = 0;
        self.timestamp = 0;
    }

    receive("test_edge_cases") {
        let a: Int = now() > 10 ? 42 : now();  //BAD: Ternary operator with now()
        self.helperFunction(now());           //BAD: Function call with now()
        self.taintedValue = self.helperFunction(now());  //BAD: Assigning tainted return value

        let nested: Int = (now() > 0 ? self.taintedValue : 0) + 
                         (self.helperFunction(now()) * 2);  //BAD: Nested tainted operations
    }

    fun helperFunction(x: Int): Int {
        return x + 1;  // Propagates taint
    }

    fun getTainted(): Int {
        return now(); //BAD: Returning tainted timestamp
    }

    receive("test_function") {
        self.timestamp = self.getTainted(); //BAD: Assigning tainted return value
    }
}
