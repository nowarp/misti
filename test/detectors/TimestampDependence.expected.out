[LOW] TimestampDependence: Assignment uses now() or a tainted variable
test/detectors/TimestampDependence.tact:11:9:
  10 |     init() {
> 11 |         self.timestamp = now();  //BAD: Assignment uses now()
               ^
  12 |         self.data = Data {
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Assignment uses now() or a tainted variable
test/detectors/TimestampDependence.tact:12:9:
  11 |         self.timestamp = now();  //BAD: Assignment uses now()
> 12 |         self.data = Data {
               ^
  13 |             time: now(),         //BAD: Struct initialization with now()
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Variable declaration uses now() or a tainted variable
test/detectors/TimestampDependence.tact:39:9:
  38 |     receive("test_variable") {
> 39 |         let time: Int = now();  //BAD: Variable declaration with now()
               ^
  40 |         self.timestamp = time; //BAD: Assigning tainted variable
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Assignment uses now() or a tainted variable
test/detectors/TimestampDependence.tact:40:9:
  39 |         let time: Int = now();  //BAD: Variable declaration with now()
> 40 |         self.timestamp = time; //BAD: Assigning tainted variable
               ^
  41 |     }
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Variable declaration uses now() or a tainted variable
test/detectors/TimestampDependence.tact:45:9:
  44 |     receive("test_loops") {
> 45 |         let start: Int = now();  //BAD: Variable initialization with now()
               ^
  46 | 
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Loop condition depends on tainted timestamp
test/detectors/TimestampDependence.tact:47:9:
  46 | 
> 47 |         while (now() < start + 1000) {  //BAD: Loop condition depends on now()
               ^
  48 |             self.timestamp += 1;
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Loop condition depends on tainted timestamp
test/detectors/TimestampDependence.tact:51:9:
  50 | 
> 51 |         repeat(5) {
               ^
  52 |             self.timestamp = now();  //BAD: Repeat loop with now()
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Assignment uses now() or a tainted variable
test/detectors/TimestampDependence.tact:52:13:
  51 |         repeat(5) {
> 52 |             self.timestamp = now();  //BAD: Repeat loop with now()
                   ^
  53 |         }
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Loop condition depends on tainted timestamp
test/detectors/TimestampDependence.tact:55:9:
  54 | 
> 55 |         do {
               ^
  56 |             self.timestamp += 1;
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Assignment uses now() or a tainted variable
test/detectors/TimestampDependence.tact:56:13:
  55 |         do {
> 56 |             self.timestamp += 1;
                   ^
  57 |         } until (now() > start + 2000);  //BAD: Until condition depends on now()
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Returning a tainted timestamp
test/detectors/TimestampDependence.tact:62:9:
  61 |     fun getTimestamp(): Int {
> 62 |         return now();  //BAD: Returning now() directly
               ^
  63 |     }
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Variable declaration uses now() or a tainted variable
test/detectors/TimestampDependence.tact:67:9:
  66 |     receive("test_expressions") {
> 67 |         let result: Int = now() > 0 ? now() : 0;  //BAD: Ternary operator with now()
               ^
  68 |         self.timestamp = result;                 //BAD: Assigning tainted result
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Assignment uses now() or a tainted variable
test/detectors/TimestampDependence.tact:68:9:
  67 |         let result: Int = now() > 0 ? now() : 0;  //BAD: Ternary operator with now()
> 68 |         self.timestamp = result;                 //BAD: Assigning tainted result
               ^
  69 |     }
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Variable declaration uses now() or a tainted variable
test/detectors/TimestampDependence.tact:82:9:
  81 |     receive("test_edge_cases") {
> 82 |         let a: Int = now() > 10 ? 42 : now();  //BAD: Ternary operator with now()
               ^
  83 |         self.helperFunction(now());           //BAD: Function call with now()
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Time-dependent usage found
test/detectors/TimestampDependence.tact:83:9:
  82 |         let a: Int = now() > 10 ? 42 : now();  //BAD: Ternary operator with now()
> 83 |         self.helperFunction(now());           //BAD: Function call with now()
               ^
  84 |         self.taintedValue = self.helperFunction(now());  //BAD: Assigning tainted return value
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Assignment uses now() or a tainted variable
test/detectors/TimestampDependence.tact:84:9:
  83 |         self.helperFunction(now());           //BAD: Function call with now()
> 84 |         self.taintedValue = self.helperFunction(now());  //BAD: Assigning tainted return value
               ^
  85 | 
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Variable declaration uses now() or a tainted variable
test/detectors/TimestampDependence.tact:86:9:
  85 | 
> 86 |         let nested: Int = (now() > 0 ? self.taintedValue : 0) + 
               ^
  87 |                          (self.helperFunction(now()) * 2);  //BAD: Nested tainted operations
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Returning a tainted timestamp
test/detectors/TimestampDependence.tact:95:9:
  94 |     fun getTainted(): Int {
> 95 |         return now(); //BAD: Returning tainted timestamp
               ^
  96 |     }
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence

[LOW] TimestampDependence: Assignment uses now() or a tainted variable
test/detectors/TimestampDependence.tact:99:9:
   98 |     receive("test_function") {
>  99 |         self.timestamp = self.getTainted(); //BAD: Assigning tainted return value
                ^
  100 |     }
Conditional logic depends on `now()`.
Help: Avoid using `now()` for logic. Replace it with deterministic or external data.
See: https://nowarp.io/tools/misti/docs/detectors/TimestampDependence